content/docs/middleware.mdx

```diff
@@ -0,0 +1,190 @@
+
+```js
+const middleware = (
+    state: JsonFormsCore,
+    action: CoreActions,
+    defaultReducer: (state: JsonFormsCore, action: CoreActions) => JsonFormsCore
+  ) => {
+    switch (action.type) {
+        case INIT: {
+            //do whatever you like
+            const newState = customStateInit();
+            return newState
+        }
+        case UPDATE_DATA: {
+            const newState = defaultReducer(state, action);
+            update(newState);
+            return newState
+        }
+        default:
+            return defaultReducer(state, action);
+    }
+  }
+...
+
+<JsonForms
+  schema={schema}
+  uischema={uischema}
+  data={data}
+  renderers={renderers}
+  onChange={({ errors, data }) => setData(data)}
+  middleware
+/>
+```
+
+### Dependent Fields
+
+
+In the following example, we'll explore a scenario where one field depends on another.
+For instance, consider a carwash service that offers various services and calculates a price based on the selected options.
+We can utilize middleware to compute and establish the price. When an `UPDATE_DATA` action is triggered, we initially invoke the default middleware to update the data and identify any errors.
+Subsequently, we adjust the price fields based on the selected services and update the state with the newly calculated data.
+
+```js
+import { INIT, UPDATE_DATA } from  '@jsonforms/core'
+
+...
+  const [data, setData] = useState(inputData);
+
+  const middleware = useCallback((state, action, defaultMiddleware) => {
+    const newState = defaultMiddleware(state,action)
+    switch (action.type) {
+        case INIT:
+        case UPDATE_DATA: {
+            if(newState.data.services.length*15 !== newState.data.price){
+              newState.data.price =  newState.data.services.length*15;
+            }
+            setData(newState.data)
+            return newState;
+        }
+    default:
+        return newState
+    }
+  });
+
+...
+
+  return (
+    <JsonForms
+      data={data}
+      schema={schema}
+      renderers={materialRenderers}
+      middleware={middleware}
+    />
+  );
+```
+
+<DependentFieldExample/>
+
+### Backend Validation
+
+In the following example, we'll look at a form that validates a secret code (1234) requiring verification by a backend service.
+Using middleware, we'll implement this functionality in JSON Forms with a controlled approach.
+The data and errors are stored in the parent components state.
+When an `INIT` or `UPDATE_DATA` action is triggered, we update the data in the parent's state, invoke our custom validation function and then return the original state. 
+The validation function asynchronously calls a function from an API-service to validate the code. If the code is invalid, an error object is set in the state and passed to the 'additionalErrors' property of JSON Forms.
```

Comment:
We should get rid of the arbitrary backend validation example. The use case should be that `data` and `errors` are managed by the parent component of JSON Forms.
Especially point to the `return state` statement as this disables internal state updates within JSON Forms. Thereby the state of JSON Forms is completely controlled from the outside.